### PR TITLE
T5681: Firewall,Nat and Nat66: simplified and standarize interface matcher

### PR DESCRIPTION
**Pull request by nicolas-fort**
## Change Summary

Use same type of commands when matching interface/interface-group in firewall, nat and nat66.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5681
* https://vyos.dev/T5643

## Component(s) name

firewall, nat, nat66
## How to test

Test configuration:
```
vyos@iface-new# run show config comm | grep "nat\|fire" | grep interf
set firewall group interface-group IF01 interface 'bond99'
set firewall ipv4 forward filter rule 10 inbound-interface name 'eth3'
set firewall ipv4 forward filter rule 10 outbound-interface group 'IF01'
set firewall ipv6 name FOO6 rule 199 inbound-interface name '!vtun11'
set firewall ipv6 name FOO6 rule 199 outbound-interface group 'IF01'
set nat destination rule 10 inbound-interface name 'eth3'
set nat source rule 20 outbound-interface group 'IF01'
set nat66 destination rule 10 inbound-interface name 'eth2'
set nat66 source rule 10 outbound-interface name 'eth3'

[edit]
vyos@iface-new#
[edit]
vyos@iface-new# sudo nft -s list table ip vyos_nat | grep "oif\|iif"
                iifname "eth3" counter dnat to 198.51.100.1 comment "DST-NAT-10"
                oifname @I_IF01 counter snat to 192.0.2.2 comment "SRC-NAT-20"
[edit]
vyos@iface-new# sudo nft -s list table ip6 vyos_nat | grep "oif\|iif"
                iifname "eth2" counter dnat to 2001:db8::1 comment "DST-NAT66-10"
                oifname "eth3" counter snat to 2001:db8::22 comment "SRC-NAT66-10"
[edit]
vyos@iface-new# sudo nft -s list table ip vyos_filter | grep "oif\|iif"
                iifname "eth3" oifname @I_IF01 counter accept comment "ipv4-FWD-filter-10"
[edit]
vyos@iface-new# sudo nft -s list table ip6 vyos_filter | grep "oif\|iif"
                iifname != "vtun11" oifname @I_IF01 counter reject comment "ipv6-NAM-FOO6-199"
[edit]
vyos@iface-new# 
```
Migration script tests:
```
## Check current config for interfaces in nat, nat66 and firewall
root@latest:/home/vyos# vyos-config-to-commands config.boot | grep "fire\|nat" | grep interface
set firewall group interface-group IG interface 'eth6'
set firewall group interface-group IG interface 'bond99'
set firewall ipv4 forward filter rule 10 inbound-interface interface-name 'vtun10'
set firewall ipv4 name FOO rule 20 inbound-interface interface-name '!eth20'
set firewall ipv4 name FOO rule 20 outbound-interface interface-group '!IG'
set firewall ipv4 output filter rule 11 outbound-interface interface-group 'IG'
set firewall ipv6 forward filter rule 1 inbound-interface interface-name 'vtun6'
set firewall ipv6 forward filter rule 1 outbound-interface interface-name '!pppoe55'
set firewall ipv6 name FOO6 rule 10 inbound-interface interface-name 'bond999'
set firewall ipv6 name FOO6 rule 20 inbound-interface interface-group '!IG'
set firewall ipv6 name FOO6 rule 20 outbound-interface interface-group 'IG'
set nat destination rule 10 inbound-interface 'eth2'
set nat source rule 10 outbound-interface 'eth3'
set nat66 destination rule 10 inbound-interface 'bri99'
set nat66 source rule 10 outbound-interface 'eth1'

root@latest:/home/vyos# ./fwall_11-to-12 config.boot
root@latest:/home/vyos# ./nat_5-to-6 config.boot
root@latest:/home/vyos# ./nat_6-to-7 config.boot
root@latest:/home/vyos# ./nat66_1-to-2 config.boot
root@latest:/home/vyos# 

## Re check that all interfaces commands were migrated.
root@latest:/home/vyos# vyos-config-to-commands config.boot | grep "fire\|nat" | grep interface
set firewall group interface-group IG interface 'eth6'
set firewall group interface-group IG interface 'bond99'
set firewall ipv4 forward filter rule 10 inbound-interface name 'vtun10'
set firewall ipv4 name FOO rule 20 inbound-interface name '!eth20'
set firewall ipv4 name FOO rule 20 outbound-interface group '!IG'
set firewall ipv4 output filter rule 11 outbound-interface group 'IG'
set firewall ipv6 forward filter rule 1 inbound-interface name 'vtun6'
set firewall ipv6 forward filter rule 1 outbound-interface name '!pppoe55'
set firewall ipv6 name FOO6 rule 10 inbound-interface name 'bond999'
set firewall ipv6 name FOO6 rule 20 inbound-interface group '!IG'
set firewall ipv6 name FOO6 rule 20 outbound-interface group 'IG'
set nat destination rule 10 inbound-interface name 'eth2'
set nat source rule 10 outbound-interface name 'eth3'
set nat66 destination rule 10 inbound-interface name 'bri99'
set nat66 source rule 10 outbound-interface name 'eth1'
root@latest:/home/vyos# 
```

## Smoketest result

```
root@iface-new:/usr/libexec/vyos/tests/smoke/cli# ./test_firewall.py 
test_bridge_basic_rules (__main__.TestFirewall.test_bridge_basic_rules) ... ok
test_flow_offload (__main__.TestFirewall.test_flow_offload) ... 
Interface "eth0" does not support hardware offload

ok
test_geoip (__main__.TestFirewall.test_geoip) ... Updating GeoIP. Please wait...
Updating GeoIP. Please wait...
ok
test_groups (__main__.TestFirewall.test_groups) ... ok
test_ipv4_advanced (__main__.TestFirewall.test_ipv4_advanced) ... ok
test_ipv4_basic_rules (__main__.TestFirewall.test_ipv4_basic_rules) ... ok
test_ipv4_mask (__main__.TestFirewall.test_ipv4_mask) ... ok
test_ipv4_state_and_status_rules (__main__.TestFirewall.test_ipv4_state_and_status_rules) ... ok
test_ipv4_synproxy (__main__.TestFirewall.test_ipv4_synproxy) ... 
"synproxy" option allowed only for action synproxy

ok
test_ipv6_advanced (__main__.TestFirewall.test_ipv6_advanced) ... ok
test_ipv6_basic_rules (__main__.TestFirewall.test_ipv6_basic_rules) ... ok
test_ipv6_mask (__main__.TestFirewall.test_ipv6_mask) ... ok
test_nested_groups (__main__.TestFirewall.test_nested_groups) ... 
Group "smoketest_network1" has a circular reference

ok
test_source_validation (__main__.TestFirewall.test_source_validation) ... ok
test_sysfs (__main__.TestFirewall.test_sysfs) ... ok
test_zone_basic (__main__.TestFirewall.test_zone_basic) ... ok

----------------------------------------------------------------------
Ran 16 tests in 37.435s

OK
root@iface-new:/usr/libexec/vyos/tests/smoke/cli# 

root@iface-new:/usr/libexec/vyos/tests/smoke/cli# ./test_nat.py 
test_dnat (__main__.TestNAT.test_dnat) ... ok
test_dnat_negated_addresses (__main__.TestNAT.test_dnat_negated_addresses) ... ok
test_dnat_redirect (__main__.TestNAT.test_dnat_redirect) ... ok
test_dnat_without_translation_address (__main__.TestNAT.test_dnat_without_translation_address) ... ok
test_nat_balance (__main__.TestNAT.test_nat_balance) ... ok
test_nat_no_rules (__main__.TestNAT.test_nat_no_rules) ... ok
test_snat (__main__.TestNAT.test_snat) ... ok
test_snat_groups (__main__.TestNAT.test_snat_groups) ... ok
test_snat_required_translation_address (__main__.TestNAT.test_snat_required_translation_address) ... 
Source NAT configuration error in rule 5: translation requires address
and/or port

ok
test_static_nat (__main__.TestNAT.test_static_nat) ... ok

----------------------------------------------------------------------
Ran 10 tests in 17.735s

OK
root@iface-new:/usr/libexec/vyos/tests/smoke/cli# 


root@iface-new:/usr/libexec/vyos/tests/smoke/cli# ./test_nat66.py 
test_destination_nat66 (__main__.TestNAT66.test_destination_nat66) ... ok
test_destination_nat66_prefix (__main__.TestNAT66.test_destination_nat66_prefix) ... ok
test_destination_nat66_protocol (__main__.TestNAT66.test_destination_nat66_protocol) ... ok
test_destination_nat66_without_translation_address (__main__.TestNAT66.test_destination_nat66_without_translation_address) ... ok
test_nat66_no_rules (__main__.TestNAT66.test_nat66_no_rules) ... ok
test_source_nat66 (__main__.TestNAT66.test_source_nat66) ... ok
test_source_nat66_address (__main__.TestNAT66.test_source_nat66_address) ... ok
test_source_nat66_protocol (__main__.TestNAT66.test_source_nat66_protocol) ... ok
test_source_nat66_required_translation_prefix (__main__.TestNAT66.test_source_nat66_required_translation_prefix) ... 
Source NAT66 configuration error in rule 5: translation address not
specified


Source NAT66 configuration error in rule 5: translation address not
specified

ok

----------------------------------------------------------------------
Ran 9 tests in 13.876s

OK
root@iface-new:/usr/libexec/vyos/tests/smoke/cli#
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/conf_mode/nat.py [hunk 1]
```diff
@@ -151,11 +151,11 @@ def verify(nat):
             err_msg = f'Source NAT configuration error in rule {rule}:'
 
             if 'outbound_interface' in config:
-                if 'interface_name' in config['outbound_interface'] and 'interface_group' in config['outbound_interface']:
-                    raise ConfigError(f'Cannot specify both interface-group and interface-name for nat source rule "{rule}"')
-                elif 'interface_name' in config['outbound_interface']:
-                    if config['outbound_interface']['interface_name'] not in 'any' and config['outbound_interface']['interface_name'] not in interfaces():
-                        Warning(f'rule "{rule}" interface "{config["outbound_interface"]["interface_name"]}" does not exist on this system')
+                if 'name' in config['outbound_interface'] and 'group' in config['outbound_interface']:
```
Comment (sever-sever) [inline src/conf_mode/nat.py:154]: Could be simplified to
```
if {'name', 'group'} <= set(config):
```
But it is not critical ;)
```
>>> config
{'name': 'eth0', 'group': 'foo'}
>>> 
>>> {'name', 'group'} <= set(config)
True
>>> 
```

**Review by sever-sever — APPROVED** [review]

**Review by sarthurdev — APPROVED** [review]

**Review by c-po — APPROVED** [review]

**Comment by sever-sever** [comment]
@MergifyIo backport sagitta

**Comment by sever-sever** [comment]
@Mergifyio backport sagitta

**Comment by sever-sever** [comment]
@Mergifyio backport sagitta

**Pull request merged**